engineapi/backup_monitor.go

```diff
@@ -96,26 +103,91 @@ func NewBackupMonitor(logger logrus.FieldLogger,
 }
 
 func (m *BackupMonitor) monitorBackups() {
+	exponentialBackoffRequired := false
+
+	// Periodically liner timer
 	wait.PollUntil(BackupMonitorSyncPeriod, func() (done bool, err error) {
-		if err := m.syncBackups(); err != nil {
-			m.logger.Errorf("Stop monitoring because of %v", err)
+		fallback, err := m.syncBackups()
+		if err != nil && !fallback {
+			m.logger.Errorf("Stop monitoring because of %v in linear timer", err)
 			m.Close()
-			return false, err
+			return false, nil
+		}
+		if fallback {
+			exponentialBackoffRequired = true
+			return true, nil
 		}
 		return false, nil
 	}, m.ctx.Done())
+
+	if !exponentialBackoffRequired {
+		return
+	}
+
+	// Exponential back off timer
+	m.logger.Info("Fallback to exponential back off timer")
+	var (
+		initBackoff   = BackupMonitorSyncPeriod
+		maxBackoff    = 10 * time.Minute
+		resetDuration = BackupMonitorMaxRetryPeriod
+		backoffFactor = 2.0
+		jitter        = 0.0
+		clock         = clock.RealClock{}
+		retryCount    = 0
+	)
+	backoffMgr := wait.NewExponentialBackoffManager(initBackoff, maxBackoff, resetDuration, backoffFactor, jitter, clock)
+	defer backoffMgr.Backoff().Stop()
+
+	ctx, cancel := context.WithTimeout(context.Background(), BackupMonitorMaxRetryPeriod)
+	defer cancel()
+
+	for {
+		select {
+		case <-backoffMgr.Backoff().C():
+			retryCount++
+			m.logger.Debugf("Within exponential back off timer retry count %d", retryCount)
+			exponentialBackoffRequired, err := m.syncBackups()
```

Comment:
This may error out, right? Then should we stop the monitor below?

Reply:
Currently, we rely on the Backup CR `status.state` in the backup controller to stop the monitor if it's error out _or_ is completed.

Reply:
Here this `monitorBackups` goroutine will exit without any auto-recovery mechanism after erroring out. 
In the controller's perspective, the monitor is still in the map but actually there is nothing periodically talking with the engine.

Reply:
Agree. It'd cause the memory leak in the [monitors](https://github.com/longhorn/longhorn-manager/blob/e89fca5206c67423dafe2c5606f30d737b1af52e/controller/backup_controller.go#L57) map.